pubsub/azurepubsub/azurepubsub.go

```diff
@@ -0,0 +1,438 @@
+// Copyright 2018 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package azurepubsub provides an implementation of pubsub that uses Azure Service Bus Topics and Subscriptions
+// PubSub.
+//
+// It exposes the following types for As:
+// Topic: *servicebus.Topic
+// Subscription: *servicebus.Subscription
+package azurepubsub // import "gocloud.dev/pubsub/azurepubsub"
+
+import (	
+	"gocloud.dev/gcerrors"
+	"context"
+	"fmt"	
+	"pack.ag/amqp"
+	"runtime"
+	"sync"
+	"time"
+	
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/driver"
+	"gocloud.dev/internal/useragent"
+
+	"github.com/Azure/azure-amqp-common-go/cbs"
+	"github.com/Azure/azure-amqp-common-go/rpc"
+	"github.com/Azure/azure-amqp-common-go/uuid"
+	"github.com/Azure/azure-service-bus-go"
+)
+
+const (
+	completedStatus  = "completed"	
+	listenerTimeout = 1 * time.Second
+	rpcTries = 5
+	rpcRetryDelay = 1 * time.Second
+)
+
+type topic struct {		
+	name string
+	ns *servicebus.Namespace	
+}
+
+// TopicOptions provides configuration options for an Azure SB Topic.
+type TopicOptions struct {}
+
+// OpenTopic opens the topic on Azure ServiceBus PubSub for the given topicName.
+func OpenTopic(ctx context.Context, topicName string, connString string, opts *TopicOptions) (*pubsub.Topic) {
```

Comment:
I was just reading the example for the [amqp package](https://godoc.org/pack.ag/amqp) you're using. It looks like there may be a way to re-use an amqp client across multiple senders and receivers.

I'm not sure if the code there is as robust or correct as the code in the actual Azure SDK, though.

Reply:
I see your point, this is why I own the sender/listener to ensure they are closed out. Problem of-course, is the tax on new amqp connection. I also wanted to keep the SB SDK types from go-cloud users. Also, what about thread safety when reusing a single sender? Should I be concerned about that?

Reply:
I wouldn't be concerned about keeping SB SDK types from our users. That's a non-goal for us. We assume there will be some provider-specific setup before getting to the OpenXXX call. (Otherwise it's just too hard to make a general API.)

You should be concerned about thread safety. Is a Sender not thread-safe?

Reply:
Yes, it is thread safe.